bite-size-notes/week-3-e-optimal-control-theory.md

```diff
@@ -0,0 +1,210 @@
+   - The objective is to minimize the total cost, which consists of both the running cost $$\sum_{k=0}^{K} J(x_k, u_k, k)$$ over the entire trajectory and the terminal cost $J_{\text{term}}(x_{K+1})$ at the final state.
+  
+#### General Setup of Optimal Control Problem
+The general Setup of Optimal Control Problem, in discrete time and finite horizon:
+
+$$\begin{aligned}
+& \min_{\begin{aligned}[t]
+& u_0, u_1, \ldots, u_K \\
+& x_0, x_1, \ldots, x_{K+1}
+\end{aligned}} && \sum_{k=0}^{K} J(x_k, u_k, k) + J_{\text{term}}(x_{K+1}) \\
+& \text{s.t.} && x_{k+1} = f(x_k,u_k,k) \quad (k = 0,1,\ldots, K) \\
+&&& x_k \in \mathcal{X}, \quad u_k \in \mathcal{U}, \quad x_0 = x_{\text{current}} \\
+&&& g_i(x_k,u_k) = 0 \quad (i = 1,\ldots, G) \\
+&&& h_i(x_k,u_k) \leq 0 \quad (i = 1, \ldots, H)
+\end{aligned}$$
+
+where $J_{\text{term}}(x_{K+1})$ is the terminal cost,
+      $$\sum_{k=0}^{K} J(x_k, u_k, k)$$ is the summation of the running cost,
+      $x_{k+1} = f(x_k,u_k,k) \quad (k = 0,1,\ldots, K)$ is the dynamics of the system,
+      $x_k \in \mathcal{X}, \quad u_k \in \mathcal{U}, \quad x_0 = x_{\text{current}}$ are the system constraints,
+      $g_i(x_k,u_k) = 0$ is an equality constraint,
+      $h_i(x_k,u_k) \leq 0$ is an inequality constraint.
+
+An additional consideration is that Optimal Control solvers require that the cost functional is smooth, so non-differentiable constraints like minimum time and obstacle avoidance must be reformulated as hard constraints, external to the cost functional. As a result, the reformulation becomes essentially an infinite-dimensional constrained optimization problem.
+
+### Solution Techniques for Optimal Control Problems
+
+There are many solution techniques available for optimal control, including, Calculus of Variations, Hamilton-Jacobi-Bellman(HJB) Equation, Direct shooting methods, etc. But for now,  we will discuss just one of these approaches in some detail, that is, Dynamic programming.
+
+#### Dynamic Programming
+
+Dynamic programming is a versatile method, especially in discrete scenarios. Take a system described by $x_{k+1} = f(x_k, u_k)$ for $k = 0, 1, \ldots, T - 1$, where $x_k$ ranges over a set $X$ of $N$ elements and $u_k$ over a set $U$ of $M$ elements. With fixed positive integers $T$, $N$, and $M$, each transition from $x_k$ to $x_{k+1}$ incurs a cost, along with a terminal cost on $X$. We aim to minimize the total cost for a trajectory up to time $T$, comprising the sum of transition costs and the terminal cost at $x_T$, given an initial state $x_0$.(See Figure 1)  
+
+The basic approach is to enumerate all possible forward trajectories from $x_0$ up to time $T$, calculate their costs, and select the optimal one. This involves $M^T$ trajectories and approximately $O(M^TT)$ operations.(See Figure 1)  
```

Comment:
"approximately $O(M^TT)$ operations" please confirm this expression with the below "$O(MT^2)$ operations for the forward scheme".

Reply:
I have corrected the typo. Thank you for pointing it out.